### PR TITLE
Support batch_size > 1 for Marian encoder-decoder models

### LINKED ISSUE #1897: Add batching support for Marian language models
**Describe the bug**
Currently, Marian models are limited to single-sequence inference. Enabling batching is critical for high-performance translation scenarios (e.g., Edge Translate), offering significant throughput improvements.

Other Models (Llama, Phi, GPT, etc.)
- **Status**: Fully Supported.
- **Details**: 
  - Models support both continuous and static batching.
  - `State` and `Model` implementations handle `[batch_size, seq_len]` inputs natively.
  - KV caching is managed per-sequence.
  - Search strategies (Greedy, Beam) operate efficiently on the full batch.

Marian Models (Current Main Branch)
- **Status**: Single Sequence Only (`batch_size = 1`).
- **Limitations**:
  - `MarianState::Run` assumes single-sequence inputs.
  - Encoder/Decoder logic does not account for padding or batch dimensions.
  - Attempting to run with `batch_size > 1` results in shape mismatches or incorrect processing of padding tokens.

**Desktop (please complete the following information):**
 - OS: Win11
 - Browser Edge
 - Version 144

**Additional context**
Testing batching performance using python script showed significant performance improvement for batch processing vs sequential translation.

#### Windows Results

| Metric | Batch | Sequential | Difference |
|--------|-------|------------|------------|
| Total Time | 1.09s | 4.76s | **4.38x slower** |
| Avg per Text | 0.011s | 0.049s | - |
| Time Wasted | - | 3.68s | 77.2% overhead |

#### Linux (WSL) Results

| Metric | Batch | Sequential | Difference |
|--------|-------|------------|------------|
| Total Time | 1.06s | 5.39s | **5.08x slower** |
| Avg per Text | 0.011s | 0.056s | - |
| Time Wasted | - | 4.33s | 80.3% overhead |

[ort_model_timing.py](https://github.com/user-attachments/files/23780152/ort_model_timing.py)

### PR DESCRIPTION
**Pull request by mustjab**
MarianInputIDs::Update assumed a single sequence: it appended EOS once to the flattened token buffer, built the encoder and decoder attention masks from next_tokens rather than the padded source, and primed the decoder by mutating next_tokens and passing a one-element subspan. Those assumptions are invisible at batch size 1 but corrupt every row beyond the first.

Append EOS per row at each sequence's unpadded length (located by scanning for pad_token_id), build both attention masks from the padded source buffer, and prime the decoder with one BOS per batch entry sized from params_->BatchBeamSize().

### CODE REVIEW

File: src/models/marian.cpp [hunk 1]
```diff
@@ -125,19 +125,46 @@ DeviceSpan<float> MarianState::Run(int current_length, DeviceSpan<int32_t>& next
 
     encoder_attention_mask_.Add();
 
-    // Create a new DeviceSpan with one additional token (0) appended
+    // Append eos per sequence at its own unpadded length. PadInputs right-pads,
+    // so appending once at the end of the flattened buffer would leave eos
+    // behind the padding for every row but the last.
+    //
+    // The prompt buffer is [batch_size, seq_len]: DefaultInputIDs and
```
[the remark this answers is not in the document]
Comment (kunal-vaishnavi) [reply src/models/marian.cpp:132]: This looks like a valid issue.

File: src/models/marian.cpp [hunk 2]
```diff
@@ -125,19 +125,46 @@ DeviceSpan<float> MarianState::Run(int current_length, DeviceSpan<int32_t>& next
...
+    // DefaultPositionInputs expand it over beams themselves, so only
+    // decoder-side tensors are sized batch*beam.
+    const size_t batch_size = static_cast<size_t>(params_->search.batch_size);
+    const int32_t pad_token_id = model_.config_->model.pad_token_id;
+    const int32_t eos_token_id = model_.config_->model.eos_token_id[0];
+
     auto original_cpu_span = next_tokens.CpuSpan();
-    auto extended_tokens = model_.p_device_inputs_->Allocate<int32_t>(original_cpu_span.size() + 1);
-    auto extended_cpu_span = extended_tokens.CpuSpan();
+    const size_t src_width = original_cpu_span.size() / batch_size;
```
Comment (kunal-vaishnavi) [inline src/models/marian.cpp:148]: From Copilot:

Could we add an explicit guard here before computing `src_width`?

Right now this assumes `batch_size > 0` and that `original_cpu_span.size()` is evenly divisible by `batch_size`. If either assumption is violated, we can end up with a divide-by-zero or silently split the flattened buffer into the wrong per-row chunks.

A small validation error would make this failure mode much easier to diagnose.

Comment (mustjab) [reply src/models/marian.cpp:148]: Done

File: src/models/marian.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (mustjab) [reply src/models/marian.cpp:132]: We don't have a Marian model included in the repo to add test for this. @kunal-vaishnavi do you know if there one available?

**Review by kunal-vaishnavi — APPROVED** [review]
